## 71% of end-end tests have passed
```
Total executed tests: 14
Failed tests: 1
Expected to fail tests: 3
Passed tests: 10
```
```
IDs of failed tests: 702745 
```
```
IDs of expected to fail tests: 741840,741839,741841 
```
<h3>Failed tests (1)</h3><details><summary>Click to expand</summary><br/><li><a href="https://ci.status.im/job/status-mobile/job/e2e/job/status-app-prs-rerun/parambuild/?BRANCH_NAME=fix/max-fee-button&PR_ID=22048&APK_NAME=22048.apk&TR_CASE_IDS=702745">Rerun failed tests</a></li><br/><h4>Class TestOneToOneChatMultipleSharedDevicesNewUi:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3811075">test_1_1_chat_non_latin_messages_stack_update_profile_photo</a>, id: 702745 </b></td></tr><tr><td><p><blockquote><div>Device 2: Wait for text element `BaseElement` to be equal to `1` </div><div>Device 2: Find `BaseElement` by `xpath`: `//*[@content-desc='activity-center-unread-count']/android.widget.TextView` </div></blockquote></p>

```
Test setup failed: critical/chats/test_1_1_public_chats.py:39: in prepare_devices
    self.home_2.handle_contact_request(self.username_1)
../views/home_view.py:383: in handle_contact_request
    self.notifications_unread_badge.wait_for_element_text(text=str(notifications_count), wait_time=120)
../views/base_element.py:172: in wait_for_element_text
    element_text = self.find_element().text.strip()
../views/base_element.py:78: in find_element
    raise NoSuchElementException(
 Device 2: BaseElement by xpath: `//*[@content-desc='activity-center-unread-count']/android.widget.TextView` is not found on the screen; For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#no-such-element-exception
```

<br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://appautomation.lambdatest.com/test?testID=APPTES4O2D4syARTVX105h">Steps, video, logs</a></li></ul></p><p>Device 2:<ul><li><a href="https://appautomation.lambdatest.com/test?testID=APPTESU074vUJYrVcOuAGO">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table></details><h3>Expected to fail tests (3)</h3><details><summary>Click to expand</summary><br/><br/><h4>Class TestWalletCollectibles:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3811086">test_wallet_send_collectible</a>, id: 741840 </b></td></tr><tr><td><p><blockquote><div>Device 1: Swiping right on element SlideButton </div><div>Device 1: Find `SlideButton` by `xpath`: `//*[@resource-id='slide-button-track']` </div></blockquote></p>

```
critical/wallet/test_collectibles.py:141: in test_wallet_send_collectible
    self.wallet_view.slide_button_track.slide()
../views/base_view.py:107: in slide
    self.swipe_right_on_element(width_percentage=1.3, start_x=100)
../views/base_element.py:288: in swipe_right_on_element
    location, size = self.get_element_coordinates()
../views/base_element.py:274: in get_element_coordinates
    element = self.find_element()
../views/base_element.py:78: in find_element
    raise NoSuchElementException(
 Device 1: SlideButton by xpath: `//*[@resource-id='slide-button-track']` is not found on the screen; For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#no-such-element-exception 
```

<b>[[Collectibles list is not loaded sometimes]]</b><br/><br/></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3811085">test_wallet_collectibles_balance</a>, id: 741839 </b></td></tr><tr><td><p><blockquote><div>Device 1: `Text` is `Double Spike` </div><div>Device 1: Click system back button </div></blockquote></p>

```
critical/wallet/test_collectibles.py:95: in test_wallet_collectibles_balance
    self.errors.verify_no_errors()
base_test_case.py:179: in verify_no_errors
    pytest.fail('\n '.join([self.errors.pop(0) for _ in range(len(self.errors))]))
 Device 1: BVL image doesn't match expected template
E    Device 1: BVL quantity 2 doesn't match expected 1
E    Device 1: BVL expanded image doesn't match expected template
E    Device 1: Glitch Punks image doesn't match expected template 
```

<b>[[Collectibles list is not loaded sometimes]]</b><br/><br/></td></tr><tr><td><b>3. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3811087">test_wallet_collectible_send_from_expanded_info_view</a>, id: 741841 </b></td></tr><tr><td><p><blockquote><div>Device 1: Find `Button` by `accessibility id`: `collectibles-tab` </div><div>Device 1: Tap on found: Button </div></blockquote></p>

```
critical/wallet/test_collectibles.py:153: in test_wallet_collectible_send_from_expanded_info_view
    self.wallet_view.get_collectible_element('Glitch Punks').wait_for_element().click()
../views/base_element.py:120: in wait_for_element
    raise TimeoutException(
 Device `1`: `CollectibleItemElement` by` xpath`: `//*[@content-desc='collectible-list-item']//*[contains(@text,'Glitch Punks')]/../..` is not found on the screen after wait_for_element 
```

<b>[[Collectibles list is not loaded sometimes]]</b><br/><br/></td></tr></tbody></table></details><h3>Passed tests (10)</h3><details><summary>Click to expand</summary><br/><br/><h4>Class TestCommunityMultipleDeviceMerged:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3811076">test_community_message_edit</a>, id: 702843 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://appautomation.lambdatest.com/test?testID=APPTESOak7jcszL24BKRMx">Steps, video, logs</a></li></ul></p><p>Device 2:<ul><li><a href="https://appautomation.lambdatest.com/test?testID=APPTESj8apBa8veqmFs2IG">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table><h4>Class TestWalletOneDevice:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3811083">test_wallet_swap_flow_mainnet</a>, id: 741555 </b></td></tr><tr><td></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3811080">test_wallet_add_remove_regular_account</a>, id: 727231 </b></td></tr><tr><td></td></tr><tr><td><b>3. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3811081">test_wallet_balance_mainnet</a>, id: 740490 </b></td></tr><tr><td></td></tr><tr><td><b>4. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3811084">test_wallet_bridge_flow_mainnet</a>, id: 741612 </b></td></tr><tr><td></td></tr><tr><td><b>5. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3811082">test_wallet_send_flow_mainnet</a>, id: 741554 </b></td></tr><tr><td></td></tr></tbody></table><h4>Class TestCommunityOneDeviceMerged:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3811074">test_community_copy_and_paste_message_in_chat_input</a>, id: 702742 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://appautomation.lambdatest.com/test?testID=APPTESiSKKTywGu8BPtHCL">Steps, video, logs</a></li></ul></p></ul></p></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3811077">test_restore_multiaccount_with_waku_backup_remove_profile_switch</a>, id: 703133 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://appautomation.lambdatest.com/test?testID=APPTESiSKKTywGu8BPtHCL">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table><h4>Class TestWalletMultipleDevice:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3811079">test_wallet_send_asset_from_drawer</a>, id: 727230 </b></td></tr><tr><td></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3811078">test_wallet_send_eth</a>, id: 727229 </b></td></tr><tr><td></td></tr></tbody></table></details>

--- a/src/quo/components/wallet/swap_input/style.cljs
+++ b/src/quo/components/wallet/swap_input/style.cljs
@@ -88,3 +88,10 @@
 
 (def gradient-end
   (assoc gradient-common :right 0))
+
+(defn button-loader
+  [theme]
+  {:width            100
+   :height           22
+   :border-radius    6
+   :background-color (loader-color theme)})

--- a/src/quo/components/wallet/swap_input/view.cljs
+++ b/src/quo/components/wallet/swap_input/view.cljs
@@ -40,6 +40,7 @@
       [:on-input-focus {:optional true} [:maybe fn?]]
       [:on-token-press {:optional true} [:maybe fn?]]
       [:on-max-press {:optional true} [:maybe fn?]]
+      [:max-loading? {:optional true} [:maybe :boolean]]
       [:customization-color {:optional true} [:maybe :schema.common/customization-color]]
       [:container-style {:optional true} [:maybe :map]]]]]
    :any])
@@ -52,7 +53,7 @@
   [{:keys [type status token value fiat-value show-approval-label? error? network-tag-props
            approval-label-props default-value auto-focus? input-disabled? enable-swap?
            currency-symbol on-change-text show-keyboard? get-ref
-           container-style on-swap-press on-token-press on-max-press on-input-focus]}]
+           container-style on-swap-press on-token-press on-max-press max-loading? on-input-focus]}]
   (let [theme                        (quo.theme/use-theme)
         pay?                         (= type :pay)
         disabled?                    (= status :disabled)
@@ -172,11 +173,13 @@
        (when-not loading?
          [:<>
           (when pay?
-            [rn/pressable {:on-press on-max-press}
-             [network-tag/view
-              (assoc network-tag-props
-                     :status
-                     (if error? :error :default))]])
+            (if max-loading?
+              [rn/view {:style (style/button-loader theme)}]
+              [rn/pressable {:on-press on-max-press}
+               [network-tag/view
+                (assoc network-tag-props
+                       :status
+                       (if error? :error :default))]]))
           (when fiat-value
             [text/text
              {:size   :paragraph-2

--- a/src/status_im/constants.cljs
+++ b/src/status_im/constants.cljs
@@ -493,10 +493,16 @@
 
 (def ^:const token-for-fees-symbol "ETH")
 
+;; The 15% buffer accounts for Ethereum's EIP-1559 fee mechanism, where the base fee can increase
+;; by up to 12.5% per block in periods of high congestion. The extra 2.5% provides additional
+;; safety to prevent failed transactions due to rapid gas price fluctuations.
+(def ^:const eth-max-fee-buffer-percent 15)
+
 (def ^:const transaction-status-success "Success")
 (def ^:const transaction-status-pending "Pending")
 (def ^:const transaction-status-failed "Failed")
 
+(def ^:const eth-send-amount-decimal 4)
 (def ^:const min-token-decimals-to-display 6)
 (def ^:const swap-proposal-refresh-interval-ms 15000)
 

--- a/src/status_im/contexts/wallet/common/utils.cljs
+++ b/src/status_im/contexts/wallet/common/utils.cljs
@@ -199,6 +199,14 @@
       (number/small-number-threshold display-decimals)
       (str amount-fixed-decimals))))
 
+(defn token-balance-for-network
+  "Returns the token balance for a specific chain"
+  [token chain-id]
+  (let [token-decimals (:decimals token)]
+    (-> (get-in token [:balances-per-chain chain-id :raw-balance] 0)
+        (number/convert-to-whole-number token-decimals)
+        money/bignumber)))
+
 (defn token-balance-display-for-network
   "Formats a token balance for a specific chain and rounds it to a specified number of decimals.
   If the balance is less than the smallest representable value based on rounding decimals,
@@ -507,3 +515,24 @@
    (-> (money/bignumber divident-price)
        (money/div (money/bignumber divisor-price))
        (number/to-fixed (min divisor-token-decimals constants/min-token-decimals-to-display)))))
+
+(defn calculate-max-safe-send-amount
+  "Calculates the max ETH that can be sent while reserving enough for gas fees.
+
+  - Ensures a minimum of 0.0001 ETH and a max of 0.01 ETH for gas.
+  - Uses 10% of the value as an estimated fee, clamped within this range.
+  - Prevents sending the full balance to avoid transaction failures.
+
+  Aligned with the desktop logic for consistency.
+  https://github.com/status-im/status-desktop/blob/f320abb5c498ac260a1a4a9db046485b88af81e7/ui/app/AppLayouts/Wallet/WalletUtils.qml#L44"
+  [value]
+  (if (or (nil? value) (zero? value))
+    "0"
+    (let [est-fee (money/maximum (money/bignumber 0.0001)
+                                 (money/minimum 0.01
+                                                (money/mul (money/bignumber value) 0.1)))
+          result  (money/sub (money/bignumber value) est-fee)]
+      (-> result
+          (money/maximum 0)
+          (number/format-decimal-fixed constants/eth-send-amount-decimal)
+          (number/remove-trailing-zeroes)))))

--- a/src/status_im/contexts/wallet/common/utils.cljs
+++ b/src/status_im/contexts/wallet/common/utils.cljs
@@ -193,7 +193,9 @@
   [amount display-decimals]
   (let [number                (or (money/bignumber amount)
                                   (money/bignumber 0))
-        amount-fixed-decimals (number/to-fixed number display-decimals)]
+        amount-fixed-decimals (-> number
+                                  (number/format-decimal-fixed display-decimals)
+                                  (number/remove-trailing-zeroes))]
     (if (and (= amount-fixed-decimals "0")
              (money/above-zero? amount))
       (number/small-number-threshold display-decimals)

--- a/src/status_im/contexts/wallet/common/utils_test.cljs
+++ b/src/status_im/contexts/wallet/common/utils_test.cljs
@@ -383,3 +383,18 @@
     (is (= (utils/calculate-token-fiat-change 50 0) 0.0))
     (is (= (utils/calculate-token-fiat-change 100 -100) 100.0))
     (is (= (utils/calculate-token-fiat-change 0.001 0.1) 0.000001))))
+
+(deftest calculate-max-safe-send-amount-test
+  (testing "Calculates the max ETH sendable while reserving fees"
+    (is (= "0.99" (utils/calculate-max-safe-send-amount "1.0")))
+    (is (= "0.009" (utils/calculate-max-safe-send-amount "0.01")))
+
+    (is (= "0" (utils/calculate-max-safe-send-amount "0.0001"))) ; Min fee is 0.0001, so 0 left
+    (is (= "0" (utils/calculate-max-safe-send-amount "0.00009"))) ; Below min fee, should return 0
+    (is (= "0.99" (utils/calculate-max-safe-send-amount "1.0")))
+
+    (is (= "9.99" (utils/calculate-max-safe-send-amount "10.0")))
+    (is (= "99.99" (utils/calculate-max-safe-send-amount "100.0")))
+
+    (is (= "0" (utils/calculate-max-safe-send-amount "0")))
+    (is (= "0" (utils/calculate-max-safe-send-amount nil)))))

--- a/src/status_im/contexts/wallet/common/utils_test.cljs
+++ b/src/status_im/contexts/wallet/common/utils_test.cljs
@@ -202,7 +202,7 @@
     (is (= (utils/sanitized-token-amount-to-display 0.0001 3) "<0.001"))
     (is (= (utils/sanitized-token-amount-to-display 0.00001 3) "<0.001"))
     (is (= (utils/sanitized-token-amount-to-display 0 2) "0"))
-    (is (= (utils/sanitized-token-amount-to-display 123.456789 4) "123.4568"))
+    (is (= (utils/sanitized-token-amount-to-display 123.456789 4) "123.4567"))
     (is (= (utils/sanitized-token-amount-to-display 0.00000123 6) "0.000001"))
     (is (= (utils/sanitized-token-amount-to-display nil 2) "0"))
     (is (= (utils/sanitized-token-amount-to-display "invalid" 2) "0"))))

--- a/src/status_im/contexts/wallet/data_store.cljs
+++ b/src/status_im/contexts/wallet/data_store.cljs
@@ -319,3 +319,40 @@
   (-> collectible
       transforms/json->clj
       transform-collectible))
+
+(defn- bridge-amount-greater-than-bonder-fees?
+  [{{token-decimals :decimals} :from-token
+    bonder-fees                :tx-bonder-fees
+    amount-in                  :amount-in}]
+  (let [bonder-fees      (utils.money/token->unit bonder-fees token-decimals)
+        amount-to-bridge (utils.money/token->unit amount-in token-decimals)]
+    (> amount-to-bridge bonder-fees)))
+
+(defn- remove-multichain-routes
+  [routes]
+  (if (> (count routes) 1)
+    [] ;; if route is multichain, we remove it
+    routes))
+
+(defn- remove-invalid-bonder-fees-routes
+  [routes]
+  (filter bridge-amount-greater-than-bonder-fees? routes))
+
+(defn- ->old-route-paths
+  [routes]
+  (map new->old-route-path routes))
+
+(def ^:private best-routes-fix
+  (comp ->old-route-paths
+        remove-invalid-bonder-fees-routes
+        remove-multichain-routes))
+
+(def ^:private candidates-fix
+  (comp ->old-route-paths remove-invalid-bonder-fees-routes))
+
+(defn fix-routes
+  [data]
+  (-> data
+      (rpc->suggested-routes)
+      (update :best best-routes-fix)
+      (update :candidates candidates-fix)))

--- a/src/status_im/contexts/wallet/send/events.cljs
+++ b/src/status_im/contexts/wallet/send/events.cljs
@@ -12,7 +12,6 @@
     [taoensso.timbre :as log]
     [utils.address]
     [utils.i18n :as i18n]
-    [utils.money :as utils.money]
     [utils.number]
     [utils.re-frame :as rf]
     [utils.security.core :as security]
@@ -626,43 +625,6 @@
                                    {:event :wallet/stop-get-suggested-routes
                                     :error error}))}]]]}))
 
-(defn- bridge-amount-greater-than-bonder-fees?
-  [{{token-decimals :decimals} :from-token
-    bonder-fees                :tx-bonder-fees
-    amount-in                  :amount-in}]
-  (let [bonder-fees      (utils.money/token->unit bonder-fees token-decimals)
-        amount-to-bridge (utils.money/token->unit amount-in token-decimals)]
-    (> amount-to-bridge bonder-fees)))
-
-(defn- remove-multichain-routes
-  [routes]
-  (if (> (count routes) 1)
-    [] ;; if route is multichain, we remove it
-    routes))
-
-(defn- remove-invalid-bonder-fees-routes
-  [routes]
-  (filter bridge-amount-greater-than-bonder-fees? routes))
-
-(defn- ->old-route-paths
-  [routes]
-  (map data-store/new->old-route-path routes))
-
-(def ^:private best-routes-fix
-  (comp ->old-route-paths
-        remove-invalid-bonder-fees-routes
-        remove-multichain-routes))
-
-(def ^:private candidates-fix
-  (comp ->old-route-paths remove-invalid-bonder-fees-routes))
-
-(defn- fix-routes
-  [data]
-  (-> data
-      (data-store/rpc->suggested-routes)
-      (update :best best-routes-fix)
-      (update :candidates candidates-fix)))
-
 (rf/reg-event-fx
  :wallet/handle-suggested-routes
  (fn [{:keys [db]} [data]]
@@ -688,8 +650,8 @@
                  (cond
                    (and failure? swap?) [:wallet/swap-proposal-error error]
                    failure?             [:wallet/suggested-routes-error error-message]
-                   swap?                [:wallet/swap-proposal-success (fix-routes data)]
-                   :else                [:wallet/suggested-routes-success (fix-routes data)
+                   swap?                [:wallet/swap-proposal-success (data-store/fix-routes data)]
+                   :else                [:wallet/suggested-routes-success (data-store/fix-routes data)
                                          enough-assets?])]]}))))))
 
 (rf/reg-event-fx

--- a/src/status_im/contexts/wallet/swap/events.cljs
+++ b/src/status_im/contexts/wallet/swap/events.cljs
@@ -2,6 +2,7 @@
   (:require [re-frame.core :as rf]
             [status-im.constants :as constants]
             [status-im.contexts.wallet.common.utils :as utils]
+            [status-im.contexts.wallet.data-store :as data-store]
             [status-im.contexts.wallet.send.utils :as send-utils]
             [status-im.contexts.wallet.sheets.network-selection.view :as network-selection]
             [status-im.contexts.wallet.swap.utils :as swap-utils]
@@ -577,3 +578,82 @@
              [:dispatch
               [:navigate-to-within-stack
                [:screen/wallet.swap-select-asset-to-pay :screen/wallet.swap-select-account]]]])})))
+
+(rf/reg-event-fx :wallet/get-swap-proposal-fee
+ (fn [{:keys [db]} [{:keys [amount-in amount-out]}]]
+   (let [wallet-address          (get-in db [:wallet :current-viewing-account-address])
+         {:keys [asset-to-pay asset-to-receive
+                 network]}       (get-in db [:wallet :ui :swap])
+         test-networks-enabled?  (get-in db [:profile/profile :test-networks-enabled?])
+         networks                ((if test-networks-enabled? :test :prod)
+                                  (get-in db [:wallet :networks]))
+         network-chain-ids       (map :chain-id networks)
+         pay-token-decimal       (:decimals asset-to-pay)
+         pay-token-id            (:symbol asset-to-pay)
+         receive-token-id        (:symbol asset-to-receive)
+         receive-token-decimals  (:decimals asset-to-receive)
+         gas-rates               constants/gas-rate-medium
+         amount-in-hex           (if amount-in
+                                   (send-utils/amount-in-hex amount-in pay-token-decimal)
+                                   0)
+         amount-out-hex          (when amount-out
+                                   (send-utils/amount-in-hex amount-out receive-token-decimals))
+         to-address              wallet-address
+         from-address            wallet-address
+         swap-chain-id           (:chain-id network)
+         disabled-to-chain-ids   (filter #(not= % swap-chain-id) network-chain-ids)
+         disabled-from-chain-ids (filter #(not= % swap-chain-id) network-chain-ids)
+         from-locked-amount      {}
+         send-type               constants/send-type-swap
+         request-uuid            (str (random-uuid))
+         params                  [(cond->
+                                    {:uuid                 request-uuid
+                                     :sendType             send-type
+                                     :addrFrom             from-address
+                                     :addrTo               to-address
+                                     :tokenID              pay-token-id
+                                     :toTokenID            receive-token-id
+                                     :disabledFromChainIDs disabled-from-chain-ids
+                                     :disabledToChainIDs   disabled-to-chain-ids
+                                     :gasFeeMode           gas-rates
+                                     :fromLockedAmount     from-locked-amount
+                                     :amountOut            (or amount-out-hex "0x0")}
+                                    amount-in (assoc :amountIn amount-in-hex))]]
+     {:db            (assoc-in db [:wallet :ui :swap :loading-swap-proposal-fee?] true)
+      :json-rpc/call [{:method     "wallet_getSuggestedRoutes"
+                       :params     params
+                       :on-success (fn [data]
+                                     (let [swap-proposal (data-store/fix-routes data)]
+                                       (rf/dispatch [:wallet/swap-proposal-fee-success
+                                                     swap-proposal])))
+                       :on-error   (fn [error]
+                                     (rf/dispatch [:wallet/swap-proposal-fee-error])
+                                     (log/error "failed to get suggested routes"
+                                                {:event  :wallet/get-swap-proposal-fee
+                                                 :error  (:message error)
+                                                 :params params}))}]})))
+
+(rf/reg-event-fx
+ :wallet/swap-proposal-fee-success
+ (fn [{:keys [db]} [swap-proposal]]
+   (let [best-routes         (:best swap-proposal)
+         selected-route      (first best-routes)
+         relevant-fee-fields [:gas-amount :gas-fees :token-fees :approval-required
+                              :approval-fee :approval-l-1-fee :bonder-fees]
+         fee-data            (select-keys selected-route relevant-fee-fields)]
+     {:db (update-in db
+                     [:wallet :ui :swap]
+                     assoc
+                     :loading-swap-proposal-fee? false
+                     :swap-proposal
+                     (when-not (empty? best-routes)
+                       fee-data))})))
+
+(rf/reg-event-fx
+ :wallet/swap-proposal-fee-error
+ (fn [{:keys [db]}]
+   {:db (update-in db
+                   [:wallet :ui :swap]
+                   assoc
+                   :loading-swap-proposal-fee?
+                   false)}))

--- a/src/status_im/contexts/wallet/swap/select_asset_to_pay/view.cljs
+++ b/src/status_im/contexts/wallet/swap/select_asset_to_pay/view.cljs
@@ -22,9 +22,12 @@
   [search-text on-change-text]
   (let [snt-token      (rf/sub [:wallet/token-by-symbol "SNT"])
         eth-token      (rf/sub [:wallet/token-by-symbol "ETH"])
+        usdc-token     (rf/sub [:wallet/token-by-symbol "USDC"])
         on-token-press (fn [token]
                          (let [pay-token-symbol (:symbol token)
-                               asset-to-receive (if (= pay-token-symbol "SNT") eth-token snt-token)]
+                               asset-to-receive (cond (= pay-token-symbol "SNT") eth-token
+                                                      (= pay-token-symbol "ETH") usdc-token
+                                                      :else                      snt-token)]
                            (rf/dispatch [:wallet.swap/start
                                          {:asset-to-pay     {:symbol pay-token-symbol}
                                           :asset-to-receive asset-to-receive

--- a/src/status_im/contexts/wallet/swap/setup_swap/view.cljs
+++ b/src/status_im/contexts/wallet/swap/setup_swap/view.cljs
@@ -182,7 +182,8 @@
                                                                available-crypto-limit
                                                                constants/eth-send-amount-decimal)
                                                               (money/to-string available-crypto-limit))]
-                                              (on-max-press max-value)))
+                                              (when (money/greater-than max-value 0)
+                                                (on-max-press max-value))))
                                           [available-crypto-limit eth-proposal?])]
     (rn/use-unmount #(rf/dispatch [:wallet/clean-swap]))
     (rn/use-effect

--- a/src/status_im/contexts/wallet/swap/setup_swap/view.cljs
+++ b/src/status_im/contexts/wallet/swap/setup_swap/view.cljs
@@ -135,7 +135,9 @@
                                                   pay-input-num-value])
         available-crypto-limit           (rf/sub [:wallet/swap-available-crypto-limit])
         display-decimals                 (min pay-token-decimals
-                                              constants/min-token-decimals-to-display)
+                                              (if eth-proposal?
+                                                constants/eth-send-amount-decimal
+                                                constants/min-token-decimals-to-display))
         available-crypto-limit-display   (utils/sanitized-token-amount-to-display
                                           available-crypto-limit
                                           display-decimals)

--- a/src/status_im/contexts/wallet/swap/setup_swap/view.cljs
+++ b/src/status_im/contexts/wallet/swap/setup_swap/view.cljs
@@ -115,6 +115,7 @@
         pay-token-symbol                 (rf/sub [:wallet/swap-asset-to-pay-symbol])
         pay-token-decimals               (rf/sub [:wallet/swap-asset-to-pay-decimals])
         loading-swap-proposal?           (rf/sub [:wallet/swap-loading-swap-proposal?])
+        loading-swap-proposal-fee?       (rf/sub [:wallet/swap-loading-swap-proposal-fee?])
         swap-proposal                    (rf/sub [:wallet/swap-proposal-without-fees])
         approval-required                (rf/sub [:wallet/swap-proposal-approval-required])
         approval-amount-required         (rf/sub [:wallet/swap-proposal-approval-amount-required])
@@ -122,6 +123,7 @@
         approval-transaction-id          (rf/sub [:wallet/swap-approval-transaction-id])
         approved-amount                  (rf/sub [:wallet/swap-approved-amount])
         error-response                   (rf/sub [:wallet/swap-error-response])
+        eth-proposal?                    (= pay-token-symbol "ETH")
         overlay-shown?                   (boolean (:sheets (rf/sub [:bottom-sheet])))
         input-ref                        (rn/use-ref-atom nil)
         set-input-ref                    (rn/use-callback (fn [ref] (reset! input-ref ref)))
@@ -131,12 +133,12 @@
                                                   (:chain-id network)])
         pay-token-fiat-value             (rf/sub [:wallet/swap-asset-to-pay-amount-in-fiat
                                                   pay-input-num-value])
-        available-crypto-limit           (money/bignumber
-                                          pay-token-balance-selected-chain)
+        available-crypto-limit           (rf/sub [:wallet/swap-available-crypto-limit])
         display-decimals                 (min pay-token-decimals
                                               constants/min-token-decimals-to-display)
-        available-crypto-limit-display   (rf/sub [:wallet/swap-asset-to-pay-balance-for-chain-ui
-                                                  (:chain-id network)])
+        available-crypto-limit-display   (utils/sanitized-token-amount-to-display
+                                          available-crypto-limit
+                                          display-decimals)
         approval-amount-required-num     (when approval-amount-required
                                            (number/to-fixed (number/hex->whole
                                                              approval-amount-required
@@ -157,14 +159,35 @@
                                                 pay-input-amount))
                                           (> pay-input-amount 0)
                                           (not pay-input-error?))
+        request-swap-proposal-fee        (rn/use-callback
+                                          (fn []
+                                            (let [safe-send-amount (utils/calculate-max-safe-send-amount
+                                                                    pay-token-balance-selected-chain)]
+                                              (rf/dispatch [:wallet/get-swap-proposal-fee
+                                                            {:amount-in safe-send-amount}])))
+                                          [pay-token-balance-selected-chain])
         request-fetch-swap-proposal      (rn/use-callback
                                           (fn []
                                             (fetch-swap-proposal
                                              {:amount                      pay-input-amount
                                               :valid-input?                valid-pay-input?
                                               :clean-approval-transaction? true}))
-                                          [pay-input-amount])]
+                                          [pay-input-amount])
+        on-max-balance-press             (rn/use-callback
+                                          (fn []
+                                            (let [max-value (if eth-proposal?
+                                                              (number/format-decimal-fixed
+                                                               available-crypto-limit
+                                                               constants/eth-send-amount-decimal)
+                                                              (money/to-string available-crypto-limit))]
+                                              (on-max-press max-value)))
+                                          [available-crypto-limit eth-proposal?])]
     (rn/use-unmount #(rf/dispatch [:wallet/clean-swap]))
+    (rn/use-effect
+     (fn []
+       (when eth-proposal?
+         (request-swap-proposal-fee)))
+     [eth-proposal?])
     (rn/use-effect
      (fn []
        (request-fetch-swap-proposal))
@@ -196,7 +219,8 @@
                               input-focused?                                    :typing
                               :else                                             :disabled)
       :on-token-press       on-token-press
-      :on-max-press         #(on-max-press (str pay-token-balance-selected-chain))
+      :on-max-press         on-max-balance-press
+      :max-loading?         loading-swap-proposal-fee?
       :on-input-focus       on-input-focus
       :value                pay-input-amount
       :fiat-value           pay-token-fiat-value

--- a/src/status_im/subs/wallet/swap.cljs
+++ b/src/status_im/subs/wallet/swap.cljs
@@ -149,6 +149,11 @@
  :-> :loading-swap-proposal?)
 
 (rf/reg-sub
+ :wallet/swap-loading-swap-proposal-fee?
+ :<- [:wallet/swap]
+ :-> :loading-swap-proposal-fee?)
+
+(rf/reg-sub
  :wallet/swap-transaction-for-signing
  :<- [:wallet/swap]
  :-> :transaction-for-signing)
@@ -383,3 +388,28 @@
  (fn [[receive-token-price currency-symbol]]
    (when receive-token-price
      (utils/fiat-formatted-for-ui currency-symbol receive-token-price))))
+
+(rf/reg-sub :wallet/max-swap-fee
+ :<- [:wallet/swap-proposal-approval-required]
+ :<- [:wallet/swap-approval-fee]
+ :<- [:wallet/swap-proposal]
+ (fn [[approval-required swap-approval-fee swap-proposal]]
+   (let [wallet-swap-proposal-fee (send-utils/full-route-gas-fee [swap-proposal])]
+     (if approval-required
+       (money/add swap-approval-fee wallet-swap-proposal-fee)
+       wallet-swap-proposal-fee))))
+
+(rf/reg-sub :wallet/swap-available-crypto-limit
+ :<- [:wallet/swap-network]
+ :<- [:wallet/swap-asset-to-pay-balance-for-chain-data]
+ :<- [:wallet/max-swap-fee]
+ :<- [:wallet/swap-asset-to-pay-symbol]
+ (fn [[network asset-to-pay-with-current-account-balance max-swap-fee pay-token-symbol]]
+   (let [pay-token-balance (utils/token-balance-for-network
+                            asset-to-pay-with-current-account-balance
+                            (:chain-id network))]
+     (if (= pay-token-symbol constants/token-for-fees-symbol)
+       (let [buffered-fee (money/mul max-swap-fee
+                                     (inc (/ constants/eth-max-fee-buffer-percent 100)))]
+         (money/sub pay-token-balance buffered-fee))
+       pay-token-balance))))

--- a/src/utils/money.cljs
+++ b/src/utils/money.cljs
@@ -246,3 +246,13 @@
 (schema/=> format-amount
   [:=> [:cat [:maybe :int]]
    [:maybe :string]])
+
+(defn maximum
+  [n1 n2]
+  (when-let [[^js bn1 ^js bn2] (->bignumbers n1 n2)]
+    (if (greater-than-or-equals bn1 bn2) bn1 bn2)))
+
+(defn minimum
+  [n1 n2]
+  (when-let [[^js bn1 ^js bn2] (->bignumbers n1 n2)]
+    (if (less-than bn1 bn2) bn1 bn2)))

